rfcs/convergence/context-values.md

```diff
@@ -0,0 +1,227 @@
+# RFC: Handling React context values in components
+
+---
+
+@layershifter @bsunderhus
+
+## Summary
+
+This RFC affects converged components and proposes multiple options detailing how to handle memoized context values for React Contexts.
+
+## Background
+
+There are cases when we need to pass some `props` or parts of `state` to child components, for this purpose we are using [React Context](https://reactjs.org/docs/context.html). To avoid unnecessary updates during components rerenders we are using either [context selector approach](https://github.com/dai-shi/use-context-selector) or [memoizing context value](https://reactjs.org/docs/context.html#contextprovider).
+
+```tsx
+// 💡 an example of memoization approach
+
+function Baz(props) {
+  const { inline, size } = props;
+  const value = React.useMemo(() => ({ inline, open }), [inline, open]);
+
+  // consumers of "SampleContext" will be notified only when "value" changes
+  return <SampleContext.Provider value={value} />;
+}
+```
+
+Currently, a few components (at least `Accordion`, `AccordionHeader`, `AccordionItem`) are defining a value for React Context on its state to use it in `render*` functions:
+
+```tsx
+// ⚠ simplified example
+
+function useBazState(props) {
+  const { inline, size } = props;
+  const [open, setOpen] = React.useState();
+
+  const contextValue = React.useMemo(() => ({ inline, open }), [inline, open]);
+  const state = {
+    inline,
+    size,
+    open,
+    // 👇 "contextValue" is a part of component's state
+    contextValue,
+  };
+
+  return state;
+}
+
+function renderBaz(state) {
+  // 👇 "value" is memoized, consumer components will not rerender without need
+  return (
+    <SampleContext.Provider value={state.contextValue}>
+      <div />
+    </SampleContext.Provider>
+  );
+}
+```
+
+## Problem statement
+
+The problem comes from multiple places:
```

Comment:
What are we addressing with this?

- what is the memory implications vs re-rendering implications ? re-rendering doesnt mean DOM re-rendering so its quite cheap operation. instead memoizing all the things creates CPU and Memory issues

Reply:
> what is the memory implications vs re-rendering implications ? re-rendering doesnt mean DOM re-rendering so its quite cheap operation. 

It's not really true. Each render of a huge tree is a complicated process that invokes React's VDOM & customer's code that can cause memory leaks (solvable) & memory bloat (sometimes not solvable).

There are two examples:
- #17521 (caused by our code, ~11% reduction in app)
- #17276 (caused by re-render, ~13% reduction in app)